NFC support for device engagement and data retrieval on Wallet Instance and Verifier App (RP)

tahirfbk opened the pull request:
This PR resolves #649, #810, #811

The proximity-flow.rst has been updated to add NFC support for proximity.

tahirfbk committed:
Proximity Flow Revision

- NFC support included for device engagement and device retrieval phases.
- revised Device retrieval using BLE

tahirfbk committed:
Merge remote-tracking branch 'upstream/versione-corrente' into NFC#649

ZAnsaroudi committed:
Fix code errors, and add puml sources

tahirfbk committed:
NFC Proximity Flow

Device Engagment and Device Retrieval Tables added

ZAnsaroudi committed:
fix: correct tables for device engagement and retrieval

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+ .. literalinclude:: ../../examples/iso-device-engagement-NFC.txt
+  :language: text
+
+**Step 2**: The verifier uses its Relying Party Instance to scan the QR code and retrieve the ``DeviceEngagement`` data from the mdoc. It SHALL select one of the transmission technologies from the ones provided in the ``DeviceEngagement`` structure.
```
```suggestion
**Step 2**: The verifier uses its Relying Party Instance to scan the QR code and retrieve the ``DeviceEngagement`` data from the mdoc. It MUST select one of the transmission technologies from the ones provided in the ``DeviceEngagement`` structure.
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+.. note::
+   For the NFC device retrieval transmission technology, the contents of the Alternative Carrier Record and Carrier Configuration Record(s) SHALL comply with [`ISO18013-5`_ #9.2.2]. For the BLE device retrieval transmission technology, the contents of the Alternative Carrier Record and Carrier Configuration Record(s) shall comply with [`ISO18013-5`_ #11.1.2].
+
+- Auxiliary Data Record MUST carry the DeviceEngagement structure from the Wallet Instance to the Relying Party Instance as part of the auxiliary NDEF record in the Handover Select message. This record has the type iso.org:18013:deviceengagement , the ID reference “mdoc”, and uses the NFC forum external type format (0x04). For each Alternative Carrier record, the Auxiliary Data Reference SHALL po ...
```
```suggestion
- Auxiliary Data Record MUST carry the DeviceEngagement structure from the Wallet Instance to the Relying Party Instance as part of the auxiliary NDEF record in the Handover Select message. This record has the type ``iso.org:18013:deviceengagement``, the ID reference “mdoc”, and uses the NFC forum external type format (``0x04``). For each Alternative Carrier record, the Auxiliary Data Reference MUST point to the NDEF record containing the ``DeviceEngagement`` Structure. 
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+
+- Auxiliary Data Record MUST carry the DeviceEngagement structure from the Wallet Instance to the Relying Party Instance as part of the auxiliary NDEF record in the Handover Select message. This record has the type iso.org:18013:deviceengagement , the ID reference “mdoc”, and uses the NFC forum external type format (0x04). For each Alternative Carrier record, the Auxiliary Data Reference SHALL po ...
+
+**Step 2**: The Relying Party Instance reads the Wallet Instance's Initial NDEF (NFC Data Exchange Format) message - , which contains a service parameter record for ``urn:nfc:sn:handover``, indicating the Wallet supports Negotiated Handover. 
```
```suggestion
**Step 2**: The Relying Party Instance reads the Wallet Instance's Initial NDEF (NFC Data Exchange Format) message, which contains a service parameter record for ``urn:nfc:sn:handover``, indicating the Wallet supports Negotiated Handover. 
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+   Use of Negotiated Handover for device engagement allows negotiation of transfer methods. For BLE, it additionally allows negotiation of keys used by the transmission layer. This provides improved user experience and enhances the security of data transmission [`ISO18013-5`_ #9.2.1].
+
+.. note::
+   Proceed only if ``DeviceEngagement`` Capabilities include ``HandoverSessionEstablishmentSupport = true``. Otherwise, skip early ``SessionEstablishment``. Early ``SessionEstablishment`` is sent via a dedicated TNEP service; the same ``SessionEstablishment`` SHALL also be sent again during data retrieval and must match—if not, the Wallet Instance terminates. If early ``SessionEstablishment`` fai ...
```
```suggestion
   Proceed only if the ``DeviceEngagement`` Capabilities include ``HandoverSessionEstablishmentSupport`` set to ``true``. Otherwise, skip the early ``SessionEstablishment``. The early ``SessionEstablishment`` is sent via a dedicated TNEP service; the same ``SessionEstablishment`` SHALL also be sent again during data retrieval and MUST match. If it does not match, the Wallet Instance terminates. If the early ``SessionEstablishment`` fails to send, proceed as normal.
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+**Step 7**: Relying Party Instance closes the TNEP service.
+
+.. note::
+   Suppose an optional ``SessionEstablishment`` message is sent during Negotiated Handover (Step 5). In that case, the Wallet Instance must verify that it matches the ``SessionEstablishment`` message received during Device Retrieval (using BLE or NFC secure channel). This verification is crucial for ensuring Session Binding.
```
```suggestion
   Suppose an optional ``SessionEstablishment`` message is sent during Negotiated Handover (Step 5). In that case, the Wallet Instance MUST verify that it matches the ``SessionEstablishment`` message received during Device Retrieval (using BLE or NFC secure channel). This verification is required to ensure a correct Session Binding.
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+    
+**Step 1**: The Wallet Instance and Relying Party Instance establish a secure BLE connection [`ISO18013-5`_ #11.1]. The Relying Party Instance (central) connects to the Wallet Instance (peripheral) using the Wallet Instance service UUID provided by DeviceEngagement, then discovers services/characteristics, and enables notifications as needed.
+
+**Step 2-5**: [Optional] Wallet Instance initiates verification by preparing to check the Relying Party’s identity via the Ident characteristic - a BLE GATT characteristic that carries an identifier value as described in [`ISO18013-5`_ #11.1.3.2]. The Wallet Instance derives the expected Ident value. It reads the Relying Party's Ident characteristic, compares it to the expected Ident, and termina ...
```
```suggestion
**Step 2-5**: [Optional] Wallet Instance initiates verification by preparing to check the Relying Party’s identity via the Ident characteristic, that is a BLE GATT characteristic that carries an identifier value as described in [`ISO18013-5`_ #11.1.3.2]. The Wallet Instance derives the expected Ident value and reads the Relying Party's Ident characteristic, comparing it to the expected Ident, and terminating the BLE connection if a mismatch may occur.
```

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+``SessionEstablishment`` over NFC
+----------------------------------
+.. note::
+   If device engagement is initiated via a QR code, the Relying Party Instance has no standardised way to signal its intent to use NFC for subsequent data transfer. This could lead to a poor user experience, as the User might not be aware that they need to use NFC. This issue is avoided when NFC is used for the device engagement, as it implicitly establishes the data transfer method [`ISO18013-5` ...
```
@RosaliaGaleano @ChiaLeo @steps88 @gaecapdtd ^

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+     - **Description**
+
+   * - **PICC**
+     - Proximity Integrated Circuit Card — the Wallet Unit acting as the NFC tag.
```
```suggestion
     - Proximity Integrated Circuit Card, implemented by Wallet Units acting as a NFC tag.
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+     - Proximity Integrated Circuit Card — the Wallet Unit acting as the NFC tag.
+
+   * - **PCD**
+     - Proximity Coupling Device — the Relying Party instance that drives `APDU` exchanges.
```
```suggestion
     - Proximity Coupling Device, impletented by Relying Party instances using `APDU` exchanges.
```

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+     - Proximity Coupling Device — the Relying Party instance that drives `APDU` exchanges.
+
+   * - **AID**
+     - Application Identifier — unique ID used in SELECT `APDU` command to open the Wallet Instance.
```
```suggestion
     - Application Identifier, unique ID used in SELECT `APDU` command to engage the Wallet Instance.
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+     - Application Identifier — unique ID used in SELECT `APDU` command to open the Wallet Instance.
+
+   * - **APDU**
+     - Application Protocol Data Unit — standard message format for communication between Relying Party Instance (PCD) and Wallet (PICC) consists of Command APDUs (e.g., `SELECT`, `ENVELOPE`, `GET RESPONSE`) from the reader and Response `APDUs` from the wallet, which include data and end with status words (`SW1/SW2`).
```
```suggestion
     - Application Protocol Data Unit, a standard message format for communication between Relying Party Instance (PCD) and Wallet (PICC) consists of Command APDUs (e.g., `SELECT`, `ENVELOPE`, `GET RESPONSE`) from the reader and Response `APDUs` from the wallet, which include data exhange and end exchange using status words (`SW1/SW2`).
```

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+     - `SW1/SW2` (Status Words) — Two-byte status code at the end of every response.  Common values: `90 00 = success`, `61 XX = more data`, `6A 82 = file/application not found`.
+
+
+**Step 1**: The Relying Party Instance (PCD) sends a SELECT `APDU` command with the Application Identifier (`AID: A0 00 00 02 48 04 00`) to open the Wallet Instance.
```
```suggestion
**Step 1**: The Relying Party Instance (PCD) sends a SELECT `APDU` command with the Application Identifier (`AID: A0 00 00 02 48 04 00`) to engage the Wallet Instance.
```

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

peppelinux commented on docs/en/proximity-flow.rst:
```diff
@@ -137,31 +395,40 @@ The Device Engagement structure MUST be CBOR encoded and have at least the follo
+
+
+   * - **DeviceRetrievalMode-NFCOptions**
+     - *(map)*. Provides options for NFC connections, including the supported role (PICC or PCD) and maximum PDU command/ response sizes. See Table 2 of `ISO18013-5`_ for the detailed mapping.
```
```suggestion
     - *(map)*. Provides options for NFC connections, including the supported role (PICC or PCD) and maximum PDU command/response sizes. See Table 2 of `ISO18013-5`_ for the detailed mapping.
```

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

peppelinux reviewed (changes requested):
very valuable work, few editorial comments and italian translation required

ZAnsaroudi commented:
> @tahirfbk test plans should be also created and theri hrefs used within the normative sections

@peppelinux We’ll update the proximity test plans to match this version and do the mapping in the normative sections.

pcerqua-ipzs committed:
feat: italian translation

pcerqua-ipzs committed:
fix: removed some rst warnings and added missing files

andmoro commented on docs/en/proximity-flow.rst:
```diff
@@ -69,56 +113,270 @@ The following figure illustrates the low-level flow compliant with ISO 18013-5 f
+.. admonition:: Box C
 
-**Step 16**: After receiving User approval, the Wallet Instance MUST retrieve the requested mdoc Digital Credentials. It then MUST prepare a `SessionData` message containing these Digital Credentials, and it MUST sign the required authentication data (as part of the mdoc authentication process, as specified in [`ISO18013-5`_ #9.1.3]). It MUST encrypt it using the established session keys before t ...
+   After receiving User approval, the Wallet Instance MUST retrieve the requested mdoc Digital Credentials. It then MUST prepare a ``SessionData`` message containing these Digital Credentials, and it MUST sign the required authentication data (as part of the mdoc authentication process, as specified in [`ISO18013-5`_ #12.4]). It MUST encrypt it using the established session keys before transmitti ...
```
```suggestion
   After receiving User approval, the Wallet Instance MUST retrieve the requested mdoc Digital Credentials. It then MUST prepare a ``SessionData`` message containing these Digital Credentials, and it MUST sign the required authentication data (as part of the mdoc authentication process, as specified in [`ISO18013-5`_ #12.4]). It MUST encrypt it using the established session keys before transmitting it to the Relying Party Instance over the secure channel. The signing ensures device binding and data integrity. The mdoc response MUST be encoded in CBOR, with its structure outlined in [`ISO18013-5`_ #10.3].
```

tahirfbk committed:
Update docs/en/proximity-flow.rst

Co-authored-by: andmoro <anmoro@fbk.eu>

Commit:
editorials: IT translation editorials & other minor editorials

andreamisuraca committed:
Added Signal Hub specs (#812)

* Added signal hub specs for AS

* Move SH specs to a new rst file

* SH (#817)

* SH_en+oas

* edits_01

* edit_02

* oas_edits_01

* json_resp_pull

* AS-edits

* sh_edits

* sh_edits_02

* data_source

* sh_edits_03

* diagram

---------

Co-authored-by: andmoro <andmoro@fbk.eu>

* Apply suggestions from code review

* Apply suggestions from code review

Co-authored-by: andmoro <anmoro@fbk.eu>

* Apply suggestions from code review

Co-authored-by: andmoro <anmoro@fbk.eu>

* Apply suggestions from code review

Co-authored-by: andmoro <anmoro@fbk.eu>

* BREAKING CHANGE: removes additional `signalType` values, pseudonymization and

* BREAKING CHANGE: removes additional `signalType` values, pseudonymization and aaplication/jwt --> application/json

* fix: new line error

* fix: new_line_2

* fix: remove_psinym_shcema

* fix: OAS_enum

* Update docs/en/signal-hub-endpoint.rst

Co-authored-by: andmoro <anmoro@fbk.eu>

* fix: enum_OAS

* Update docs/en/authentic-source-endpoint.rst

* Update docs/en/authentic-source-endpoint.rst

* Update docs/en/authentic-source-endpoint.rst

* Update docs/en/signal-hub-endpoint.rst

Co-authored-by: andmoro <anmoro@fbk.eu>

* Update docs/en/signal-hub-endpoint.rst

* Update docs/en/signal-hub-endpoint.rst

* Update docs/en/signal-hub-endpoint.rst

* Apply suggestions from code review

fix: editorials

Co-authored-by: mbasili-ipzs <m.basili@ipzs.it>

* chore: SH_link_reframe

* Apply suggestions from code review

editorials: code_review

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>
Co-authored-by: mbasili-ipzs <m.basili@ipzs.it>

* Apply suggestions from code review

editorials: fixed last editorials from the code review

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

* Update docs/en/signal-hub-endpoint.rst

* feat: italian translation to be reviewed

* fix: wrong references

* fix: reverted defined terms

* Apply suggestions from code review

fix: code review editorials

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

* editorial: AS normative language

* editorials: IT translation edits

* edits: fixed note indentation

* fix: script error

* editorials: fix IT translation

* Update docs/en/signal-hub-endpoint.rst

Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>

* Apply suggestions from code review

Co-authored-by: andmoro <anmoro@fbk.eu>

* Apply suggestions from code review

fix: AS normative use of signal hub

---------

Co-authored-by: andmoro <anmoro@fbk.eu>
Co-authored-by: andmoro <andmoro@fbk.eu>
Co-authored-by: Giuseppe De Marco <gi.demarco@innovazione.gov.it>
Co-authored-by: Giuseppe De Marco <demarcog83@gmail.com>
Co-authored-by: mbasili-ipzs <m.basili@ipzs.it>
Co-authored-by: Pasquale Cerqua <p.cerqua@ipzs.it>

tahirfbk committed:
Merge remote-tracking branch 'upstream/versione-corrente' into NFC#649

ZAnsaroudi committed:
Merge branch 'rp-req-tp1' into NFC#649

ZAnsaroudi committed:
fix: align italian translations
